### ISSUE TITLE
thesis-central and thesis-central-qa are not accessible

### ISSUE THREAD

**Issue by jrgriffiniii** [pulibrary/drds_catchall#93]
NB: This MUST be done before April 16 or we will have another service interruption.

Incident response document here: https://docs.google.com/document/d/1zB9TqthyZEiFZy0hZgwi-s4J9SCyg4u5vaWN7b6zX_o/edit#

Next steps: 
1. OIT is going to spin up two EC2 instances that can support Postgres 9.6 and that can be reached from thesis-central servers
2. @jrgriffiniii and @bess will import database backups into these and ensure they work
3. @jrgriffiniii and @bess will then set up a maintenance window to snapshot latest thesis-server data and switch over to new database solution.

**Comment by jrgriffiniii** [comment]
I could not immediately diagnose the errors with these deployments. I am trying to restart the applications currently, but I keep encountering the following:

```bash
[thesiscentralsu@libtc201l Vireo]$ :: loading settings :: url = jar:file:/home/thesiscentralsu/play1-1.3.0/framework/lib/ivy-2.3.0.jar!/org/apache/ivy/core/sett
ings/ivysettings.xml
Exception in thread "main" play.exceptions.DatabaseException: Cannot connected t
o the database[default], Connections could not be acquired from the underlying d
atabase!
        at play.db.DBPlugin.onApplicationStart(DBPlugin.java:196)
        at play.plugins.PluginCollection.onApplicationStart(PluginCollection.jav
a:515)
        at play.Play.start(Play.java:537)
        at play.Play.init(Play.java:309)
        at play.server.Server.main(Server.java:160)
Caused by: java.sql.SQLException: Connections could not be acquired from the und
erlying database!
        at com.mchange.v2.sql.SqlUtils.toSQLException(SqlUtils.java:118)
        at play.db.DBPlugin.onApplicationStart(DBPlugin.java:179)
        ... 4 more
Caused by: com.mchange.v2.resourcepool.CannotAcquireResourceException: A Resourc
ePool could not acquire a resource from its primary factory or source.
        at com.mchange.v2.resourcepool.BasicResourcePool.awaitAvailable(BasicRes
ourcePool.java:1463)
```

**Comment by jrgriffiniii** [comment]
I truly do not understand why the connection is failing from inside of the application:

```bash
[thesiscentralsu@libtc201l Vireo]$ psql -h theses-central-2020.ct9hnzmkuoct.us-e
ast-1.rds.amazonaws.com -W -U thesis-prod thesis-prod-db
Password for user thesis-prod:
psql (9.6.20, server 12.9)
WARNING: psql major version 9.6, server major version 12.
         Some psql features might not work.
SSL connection (protocol: TLSv1.2, cipher: ECDHE-RSA-AES256-GCM-SHA384, bits: 256, compression: off)
Type "help" for help.

thesis-prod-db=> \q
```

**Comment by jrgriffiniii** [comment]
The database URL and credentials are being properly passed to the application here:

https://github.com/playframework/play1/blob/629aee531fd3290113bedff97f7d762646304ed9/framework/pym/play/application.py#L29

I have been checking this within the stack traces, and my fear is that this may be related to an out-of-date Java dependency.

**Comment by kelynch** [comment]
This needs to be addressed before April 16 when the next forced upgrade will take place.  Currently blocked/on hold until we have access to the EC2 instances from OIT where we can set up Postgres, import data, and test connectivity.

**Comment by jrgriffiniii** [comment]
I can confirm that I currently do have access to the new VMs `libtc203l` and `libtc103l` using the PuTTY on the bastion host `lib200w`. I am, however, limited in my user account privileges:

```bash
[libvijrg@libtc103l ~]$ sudo yum install tmux

We trust you have received the usual lecture from the local System
Administrator. It usually boils down to these three things:

    #1) Respect the privacy of others.
    #2) Think before you type.
    #3) With great power comes great responsibility.

Duo two-factor login for libvijrg

Enter a passcode or select one of the following options:

 1. Duo Push to XXX-XXX-8440
 2. Phone call to XXX-XXX-8440
 3. SMS passcodes to XXX-XXX-8440 (next code starts with: 2)

Passcode or option (1-3): 1

Pushed a login request to your device...
Success. Logging you in...
Sorry, user libvijrg is not allowed to execute '/usr/bin/yum install tmux' as root on libtc103l.aws.Princeton.EDU.
```

**Comment by jrgriffiniii** [comment]
After attempting other possible approaches to elevating my privileges, I have explicitly requested the necessary privileges by e-mail.

**Comment by jrgriffiniii** [comment]
Related to the service outage, an administrative user has provided the following error report from a user:

> An issue has come up with the files that were submitted prior to the site outage on Sunday. I received an error message when I try to view/download a file that was submitted earlier than 4/9.

 

> A student alerted me to the issue:

> 10:26

> Hello,I recently (on Saturday) uploaded my thesis onto Thesis Central. While everything on the site worked correctly then, today when I try to view the Manuscript in PDF of my thesis on the site an error occurs; in addition, when I view the submission of my thesis it now says the pdf has a size of 0 bytes. I just wanted to make you aware of the issue so that I can assure that my thesis is still properly stored on the site as it was on Saturday, especially considering that it is due today (Tuesday) at 4:30 PM.Thank you!

**Comment by jrgriffiniii** [comment]
The following postgreSQL query retrieves the paths and names for the files uploaded:

```bash
thesis-prod-db=> select a.name,a.data,sub.id from submission as sub inner join attachment as a on a.submission_id=sub.id inner join person as p on sub.submitter_id=p.id where email like '%[SUBMITTER_NAME]%';
```

**Comment by jrgriffiniii** [comment]
There clearly are files which were successfully uploaded throughout 04/10/23, I am still struggling to identify why there might be some which are missing:

```bash
[thesiscentralsu@libtc201l attachments]$ pwd
/home/thesiscentralsu/Vireo/data/attachments
[thesiscentralsu@libtc201l attachments]$ find . -type f -exec ls -l {} \;
-rw-r--r-- 1 thesiscentralsu thesiscentralsu 395858 Apr 10 11:54 ./aa/ec/00/0f/aaec000f-1641-4703-9f2d-1283bbb06c63
-rw-r--r-- 1 thesiscentralsu thesiscentralsu 2343669 Apr 10 23:05 ./aa/a2/6f/b7/aaa26fb7-1a31-4b0d-887a-e3e4b0cacce8
-rw-r--r-- 1 thesiscentralsu thesiscentralsu 248 Apr 10 13:29 ./94/42/72/0d/9442720d-202a-451f-bd89-3c7371c92ede
-rw-r--r-- 1 thesiscentralsu thesiscentralsu 250 Apr 10 15:53 ./32/41/83/76/32418376-5c42-42c6-a863-a16edb2b4c4e
-rw-r--r-- 1 thesiscentralsu thesiscentralsu 1622191 Apr 10 11:40 ./32/41/b7/49/3241b749-a356-4368-93d2-6ec92d6129da
```

**Comment by bess** [comment]
Increasing the points on this one to account for the fact that 6+ people spent much of a day on restoring missing PDFs.

**Comment by jrgriffiniii** [comment]
While this session is still in progress, it should be noted that there were a number of PDF files which were deleted during the troubleshooting for the application. In response to this, OIT has, fortunately, provided snapshots for the file system for the EC2 instances.

**Comment by kelynch** [comment]
Run book for restore process - https://docs.google.com/document/d/17dXwE9EN9GbAq7gBpQ6SgEAs4wmHs6KytCDXJZ6Q5vg/edit

**Comment by jrgriffiniii** [comment]
Currently, I am blocked from transferring the database export file from the production application server into the new QA database server environment, likely by firewall restrictions. I have issued a request for support in order to advance this:

> I am sorry for continuing this thread, however, I have reached a point at which I have deployed the postgresql server on libtc103l (for the new QA database), and I am currently trying to migrate the database export from the current production server (libtc201l for the production application). Unfortunately, I do not believe that I can achieve this over the network due to firewall restrictions.

> postgres@libtc201l$ psql -h libtc103l.aws.princeton.edu -W -U thesis-qa thesis-qa-db <  thesis_prod_db_04_11_23b.sql

> Might this be possible to please request? Otherwise, might there be some other means to transfer the database export file from libtc201l to libtc103l?

**Comment by jrgriffiniii** [comment]
Following a pairing session between @kayiwa and I, a database export from a snapshot of the current production server (libtc201l) has been successfully imported into the new PostgreSQL database server for the QA environments (libtc103l).

While there are still questions and concerns regarding how best to manage the database backup exports and restoring these backups, please know that that, once this is determined, the path for upgrading the production server environment with the new database server (libtc203l) shall be fully identified.

**Comment by jrgriffiniii** [comment]
The `pg_back` utility has been successfully installed and configured to synchronously perform backups for the production and QA PostgreSQL server. These are also uploaded to a dedicated Google Cloud Services Bucket.

**Comment by kelynch** [comment]
This was successfully addressed this past Friday with a cutover to EC2 for database.  The AWS maintenance window on Sunday passed without incident.
